drivers: dma: intel-adsp-hda: Correct DGCS:SCS bit for 32bit sample size

ujfalusi opened the pull request:
If the channel was used for 16bit in the once, subsequent 32bit sample size audio will be broken since the SCS bit remains set.

Example sequence with SOF:
normal audio playback with 16bit
ChainDMA audio playback with 16bit
normal audio playback with 16bit

The last playback results garbled audio.

Introduce intel_adsp_hda_set_sample_container_size() helper function to handle the SCS bit and use it in the driver.

ujfalusi committed:
drivers: dma: intel-adsp-hda: Correct DGCS:SCS bit for 32bit sample size

If the channel was used for 16bit in the once, subsequent 32bit sample size
audio will be broken since the SCS bit remains set.

Example sequence with SOF:
normal audio playback with 16bit
ChainDMA audio playback with 16bit
normal audio playback with 16bit

The last playback results garbled audio.

Introduce intel_adsp_hda_set_sample_container_size() helper function
to handle the SCS bit and use it in the driver.


Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

kv2019i reviewed (dismissed):
Excellent find @ujfalusi ! 

This fixes SOF bug tracked as https://github.com/thesofproject/sof/issues/8236

jsarha reviewed (approved):
Makes perfect sense.

mengdonglin commented:
> If the channel was used for 16bit in the once, subsequent 32bit sample size audio will be broken since the SCS bit remains set.
> 
> Example sequence with SOF: 
> normal audio playback with 16bit 
> ChainDMA audio playback with 16bit 
> normal audio playback with 16bit
> The last playback results garbled audio.

@ujfalusi I have silly question. The last normal playback is still 16bit, why it results garbled audio? Or does chain DMA uses 16bit container, and normal playback use 32bit container with 16bit valid?

fabiobaltieri commented on drivers/dma/dma_intel_adsp_hda.c:
```diff
@@ -53,10 +53,15 @@ int intel_adsp_hda_dma_host_in_config(const struct device *dev,
+		if (dma_cfg->source_data_size <= 3)
 			*DGCS(cfg->base, cfg->regblock_size, channel) |= DGCS_SCS;
-		}
+		else
+			*DGCS(cfg->base, cfg->regblock_size, channel) &= ~DGCS_SCS;
```
the coding style requires braces around all conditions, same in the other ones

ujfalusi replied:
Oh, I though the braces were there to not look silly with the comment, I'll add them in a minute

fabiobaltieri replied:
I know, sorry, it's one of the gaps from the Linux codingstyle.

ujfalusi replied:
no problem, I will remember from now on, thanks @fabiobaltieri for the reminder!

ujfalusi commented:
Changes since v1:
- Braces added around the if () else statements

plbossart commented on drivers/dma/dma_intel_adsp_hda.c:
```diff
@@ -121,9 +133,15 @@ int intel_adsp_hda_dma_link_in_config(const struct device *dev,
+	if (dma_cfg->source_data_size <= 3) {
 		*DGCS(cfg->base, cfg->regblock_size, channel) |= DGCS_SCS;
+	} else {
+		*DGCS(cfg->base, cfg->regblock_size, channel) &= ~DGCS_SCS;
```
this one changes the behavior if res != 0, is this intentional @ujfalusi?

ujfalusi replied:
Oh, I have missed that, sorry I will update as soon as I can.

plbossart commented on drivers/dma/dma_intel_adsp_hda.c:
```diff
@@ -153,9 +171,15 @@ int intel_adsp_hda_dma_link_out_config(const struct device *dev,
+	if (dma_cfg->source_data_size <= 3) {
 		*DGCS(cfg->base, cfg->regblock_size, channel) |= DGCS_SCS;
+	} else {
+		*DGCS(cfg->base, cfg->regblock_size, channel) &= ~DGCS_SCS;
```
same here, change with res!=0?

plbossart reviewed (commented):
The host side is fine, the link has a change that's not very clear @ujfalusi

teburd reviewed (changes requested):
Missed the ret != 0 drop that @plbossart found while looking at this on my phone screen. This does need fixing before this should be merged, so marked -1

ujfalusi commented:
Changes since v2:
- do not ignore the `res` from intel_adsp_hda_set_buffer() in the linkDMA configurations

ujfalusi commented:
One more try, sorry, I missed something else

ujfalusi commented:
Changes since v3:
- corrected the copy-paste issue of replacing dest_data_size with source_data_size.

ujfalusi commented:
Changes since v4:
- a bit more elegant way is to introduce `intel_adsp_hda_set_sample_container_size()` and use that in the driver
- the resulting code is easier to read and shorter.

ranj063 commented on soc/xtensa/intel_adsp/common/include/intel_adsp_hda.h:
```diff
@@ -469,4 +469,26 @@ static inline uint32_t intel_adsp_hda_check_buffer_interrupt(uint32_t base, uint
+		*DGCS(base, regblock_size, sid) |= DGCS_SCS;
+	} else {
+		*DGCS(base, regblock_size, sid) &= ~DGCS_SCS;
+	}
```
do we need braces for the if/else?

kv2019i replied:
@ranj063 yes, this is actually required in Zephyr
